### ISSUE TITLE
Revert "Add nested wildcards to children of abs nodes." [phab:D1366]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1763]
_Created by **benjyw** at 2017-02-23 02:32:21_: 

This reverts commit b4b5b56b818997ce62a4fe85b2307ce1d6929343.

The author of that commit provides a rationale for why it
should be reverted here:
#1562

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-02-23 02:34:45_:

  - Add Misha back to AUTHORS and CONTRIBUTORS.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2017-03-01 00:30:18_:

Sorry for the delay in getting back to you here.

If the logic should be reversed, I'd prefer a new change that reverses the logic, but keeps (and modifies) the tests.

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-03-01 05:33:30_:

Restore the test (appropriately modified).

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-03-01 05:33:45_:

Good idea. Restored the tests.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2017-03-01 22:55:13_:

I think it makes sense to replace this with

`//- !{ NestedWildFnAbs param.1 _ }`

to show that there is no param 1. There might be a better way to show that a parameter edge does not exist. zarko would know for sure.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2017-03-01 22:55:13_:

This seems reasonable to me. I'll give zarko the final say since he approved the original diff so I might be missing something obvious.

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-03-02 00:51:16_:

  - More robust test.

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-03-02 00:51:44_:

Modified the test as you suggest.

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-03-10 01:44:42_:

<gentle ping>

**Comment by kythe** [comment]
_**zarko** wrote at 2017-03-11 00:57:01_:

Sorry, was OOO.

The reason I believe the change being reverted here is correct is that I read
```
private static void nestedWildcard(Optional<? extends Generic<?>> ogvar) {}
```
as
```
\forall T, \forall S : S <: Generic<T>, nestedWildcard : Optional<S> -> ()
```
Is this incorrect?

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-03-12 06:50:06_:

This comes down to our (possibly inaccurate, please weigh in on this...) understanding of what `ABS` and `ABSVAR` nodes represent. Our current understanding (as opposed to our incomplete understanding when we authored the original change) is that the ABSVARs represent the instantiation choices required in order to instantiate an abstract subgraph into a concrete one.

In this example, `nestedWildcard` is abstract in `S` only, in the sense that once `S` has been chosen (subject to its own constraints) there are no more degrees of freedom, and we have a concrete instantiation of `nestedWildcard`.  It may so happen that the choice of `S` has its own constraints (and those are expressed in the subgraph representing the ABSVAR), but these constraints imply _less_ freedom in the choice of `S`, not more, so it doesn't seem right that there should be more ABSVARs in this case.

In short, it seems right to say that `nestedWildcard` is parameterized in `S`, but not in `T`.

Does that make sense?

**Comment by kythe** [comment]
_**zarko** wrote at 2017-03-14 22:53:08_:

The problem with dropping `T` is that you're left without the ability to express the constraint `S <: Generic<T>` (`TyvarS bounded/upper TAppGenericT`). In the test case as amended by this diff, the type variable for `T` has no binding site, so the graph is not valid.

**Comment by kythe** [comment]
_**misha** wrote at 2017-03-15 22:58:29_:

>>! In (earlier comment), zarko wrote:
> The problem with dropping `T` is that you're left without the ability to express the constraint `S <: Generic<T>` (`TyvarS bounded/upper TAppGenericT`). In the test case as amended by this diff, the type variable for `T` has no binding site, so the graph is not valid.

I apologize in advance in case I misunderstood the situation or your concern, but don't the following statements from the test assert that the constraint is expressed?

```
  //- FirstWildcard bounded/upper GType
  //- GType.node/kind tapp
  //- GType param.0 GenericClass
  //- GType param.1 SecondWildcard
```

**Comment by kythe** [comment]
_**zarko** wrote at 2017-03-15 23:08:56_:

`GType param.1 SecondWildcard` is a use of `SecondWildcard`, but because `SecondWildcard` is not a param of any `abs`, it's an illicit free variable.

**Comment by kythe** [comment]
_**misha** wrote at 2017-03-15 23:36:36_:

Ok, so if I understand correctly, the problem is not that the `abs` node needs to have `SecondWildcard` as a parameter. Instead the problem is that something needs to reference `SecondWildcard`, so that it's not a free variable. Is that right?

**Comment by kythe** [comment]
_**zarko** wrote at 2017-03-15 23:47:37_:

Not quite. We model polymorphism in the graph using `abs` and `tyvar` semantic nodes. It must be the case that, for a given `tyvar`, there exists a `param` edge directed at it from an `abs`. Otherwise what's recorded in the graph is a dangling type variable that has never been bound anywhere. This relationship is distinct from `defines/binding` or `defines` because there isn't necessarily any language-level syntax around to suggest it.

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-03-20 09:29:32_:

I've spent a good while staring at this thread and reading the Java language spec, and I think I have a better understanding of the issue now. 

A wildcard is a type //argument//, not a type parameter. See, e.g., https://docs.oracle.com/javase/specs/jls/se8/html/jls-4.html#jls-4.5.1. This should have been obvious to me from the fact that a variable can be of type Generic<?>, and there's no such thing as a "generic variable".

So `nestedWildcard` in that example is not a generic method, it is a concrete method. And if I understand the Kythe schema correctly, this means that it should not define an ABS node, and the wildcards are not ABSVARs. So this change doesn't go far enough... 

My conclusion is that a wildcard should be represented as a simple (unique) type node.

Thoughts?

**Comment by kythe** [comment]
_**zarko** wrote at 2017-03-21 22:40:21_:

From my reading of the spec:
```
private static void nestedWildcard(Optional<? extends Generic<?>> ogvar) {}
```
is not
```
\forall T, \forall S : S <: Generic<T>, nestedWildcard : Optional<S> -> ()
```
but instead:
```
Togvar = \exists T. \exists S <: Generic<T>
nestedWildcard : Togvar -> ()
```
where the existential binders are part of the type of the parameter, not of the function. We therefore need to move some binders around and decide on what to do with existentials (I vote to reuse our multipurpose binder `abs` and mark type variables with existential facts).

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-03-22 23:24:44_:

I think that's correct. A wildcard is an existential type.

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-03-23 00:15:49_:

Or, more precisely, a type with a wildcard as a type argument is an existential type.

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-03-23 01:34:47_:

Meanwhile I will abandon this change, as a much more significant one is merited.

**Comment by kythe** [comment]
_**benjyw** abandoned revision at 2017-03-23 01:34:57_

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-03-23 05:14:54_:

So a straw man proposal is: 

  //- Optional ref OptionalClass
  //- wildcardField defines/binding WF
  //- WF typed WFType
  //- WF.node/kind tapp
  //- WF param.0 OptionalClass
  //- WF param.1 Wildcard
  //- Wildcard.node/kind wildcard
  //- Number ref List
  //- Wildcard bounded/upper NumberType
  private Optional<? extends Number> wildcardField;

Where we introduce a node kind `wildcard` (or `existential`, or whatever we want to call it) and allow it to have `bounded/upper`/`bounded/lower` edges.

 This doesn't involve abs/absvar nodes, but I'm not sure they are called for?  `Optional<?>` is a single concrete type ("the type of all objects O for which there exists a type T(O) such that the object is of type Optional<T>", aka "the type of all Optionals"), so this is not a case of polymorphism, really.
 
Thoughts?

(There is a wrinkle here, namely that we'd want all instances of `? extends Foo` to have the same signature, because they do all indeed represent the same type. This means generating the signature using only the bounds.)

**Comment by kythe** [comment]
_**zarko** wrote at 2017-03-23 17:02:12_:

This is a "case of polymorphism", although you have existential quantification versus universal quantification. You still have a binder, and that binder's ∃. `Optional<? extends Foo>` means `∃ x : x <: Foo, Optional<x>`. Note that you may bind multiple variables, as in `Pair<? extends Foo, ? extends Bar>`. It's sensible that there should be one representation for this in the graph, but we're in sort of unfortunate situation where we name all of our type variables uniquely (in the Barendregt style). This is nice because it means that tools don't have to understand binding to show that T and S are distinct when users are mousing around in
```
void <T> foo(T x) { }
void <S> bar(S y) { }
```
but this falls over when we want a single representation of `∃ x, y : x <: Foo ∧ y <: Bar, Pair<x, y>`, since every time this type is uttered `x` and `y` have different identities. One possible alternate representation would use de Bruijn indices (where a variable is named after its distance from its binder), e.g.: `∃ <: Foo ∃ <: Bar, Pair<1, 0>`.

Early on we resisted this kind of representation because it's difficult for clients to work with and for indexers to generate without actually understanding binding.

Can you say anything about what you're trying to do with the graph? Maybe we can come up with a way to give you the data you need without overhauling our notion of binding (or doing something that isn't cromulent wrt types).

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-03-23 21:53:17_:

Well, when I say "not a case of polymorphism", I mean that universal quantification provides parametric polymorphism, while existential quantification does not (to the best of my understanding and recollection, it's been a long time since university...)

But this may be beside the point, depending on what `abs`/`absvar` are intended to represent...  I'm not sure in what sense an existential quantifier "abstracts over a subgraph". 

Can you modify the straw man I posted above with the assertions you had in mind? In particular, I'm not sure what anchor defines the type.


A use case we have in mind is the ability to reason about API compatibility.  For example, given these methods:
```
int foo(Optional<? extends Number> x)

Optional<? extends Number> bar(String s)
```

We'd like to be able to determine by inspecting the graph that
```
int y = foo(bar("Hello"));
```
is valid.


One thing I assume we agree on is that the current state (regardless of this change) is broken. For example, for a variable of type `Optional<?>` we currently emit a dangling absvar.

**Comment by kythe** [comment]
_**zarko** wrote at 2017-03-23 22:32:16_:

Yes, I agree with you that we're not emitting the right graph right now.

For the query you have in mind, would you be willing to crawl around in the graph a bit? For example, your tool would need to perform the following task, stated below in two ways:

 - show that two representations of `Optional<? extends Number>` are equivalent up to variable renaming
 - show that the two subgraphs for `Optional<? extends Number>` rooted at the same `abs` node are isomorphic modulo VNames for absvars

This is easy enough if `foo` and `bar` are fixed, but might need extra cleverness if you're doing something more global.

Once we figure that out, we can start thinking about representation.

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-03-23 23:30:10_:

Well, `foo` and `bar` aren't fixed, but we'd already have to do some "clever" processing, regardless of wildcards, because in the general case the type of the return value of `foo` need only be compatible with the type of the argument of `bar`,  not identical to it, so it's not like we'd just be looking at type node identity anyway.

But I'm still not sure I understand what these isomorphic subgraphs would look like: what subgraph is being abstracted over, or how it relates to the existing universally-quantified `abs` node for the generic type. Can you give an example? I'm just not sure that existential types (especially in their limited form as Java wildcards) are appropriately modeled with an `abs` node.
